vm/src/object/ext.rs

```diff
@@ -212,6 +212,49 @@ impl<T: PyPayload> ToPyObject for PyRefExact<T> {
     }
 }
 
+pub struct PyAtomicRef<T: PyObjectPayload>(AtomicCell<NonNull<Py<T>>>);
+
+cfg_if::cfg_if! {
+    if #[cfg(feature = "threading")] {
+        unsafe impl<T: Send + PyObjectPayload> Send for PyAtomicRef<T> {}
+        unsafe impl<T: Sync + PyObjectPayload> Sync for PyAtomicRef<T> {}
+    }
+}
+
+impl<T: PyObjectPayload> From<PyRef<T>> for PyAtomicRef<T> {
+    fn from(pyref: PyRef<T>) -> Self {
+        let py = PyRef::leak(pyref);
+        Self(AtomicCell::new(NonNull::from(py)))
+    }
+}
+
+impl<T: PyObjectPayload> Deref for PyAtomicRef<T> {
+    type Target = Py<T>;
+
+    fn deref(&self) -> &Self::Target {
+        unsafe { self.0.load().as_ref() }
+    }
+}
+
+impl<T: PyObjectPayload> PyAtomicRef<T> {
+    /// # Safety
+    /// The caller is responsible to keep the returned PyRef alive
+    /// until no more reference can be used via PyAtomicRef::deref()
+    #[must_use]
+    pub unsafe fn swap(&self, pyref: PyRef<T>) -> PyRef<T> {
+        let py = PyRef::leak(pyref);
+        let old = self.0.swap(NonNull::from(py));
+        PyRef::from_raw(old.as_ptr())
+    }
+
+    pub fn swap_to_temporary_refs(&self, pyref: PyRef<T>, vm: &VirtualMachine) {
+        let old = unsafe { self.swap(pyref) };
+        if let Some(frame) = vm.current_frame() {
+            frame.temporary_refs.lock().push(old.into());
```

Comment:
Does temporary_refs ever get cleared? Like I guess it doesn't need to be, necessarily, but I don't like the idea of
```py
class A: pass
class B: pass
B().__class__ = A
```
meaning that A will never ever ever be deallocated.

What if instead `class()` returned a guard.... no that wouldn't work. I'll try and think of something by the morning, maybe.

Reply:
I guess when the frame is popped

Reply:
Ohhh, I missed that its per-frame, yeah I guess that does make sense. Nevermind!

Reply:
> I guess when the frame is popped

Yes, we can clear the refs for each instruction, but I don't think it is worth.